### DISCUSSION TITLE
Export plot as image

### DISCUSSION THREAD

**Discussion by ixxie** [uwdata/mosaic#206] [Ideas]
Plenty of charting libraries offer a button that lets you download the chart in its current state as an image.

As far as I can tell this feature is not available in Mosaic at the moment.

**Comment by domoritz** [comment]
If you think of mosaic as the overall architecture or the middleware, it's agnostic to the rendering library which would be responsible for image exports. So the answer will be a bite more nuanced depending on the charting library you use. 

vgplot uses a combination of standard widgets, tables, and observable plot. You can use something like https://nytimes.github.io/svg-crowbar/ to get the images for the SVG plots but in general you are right that we down support export. 

We do have a demo of using mosaic with Vega where you can then use the Vega export.

**Reply by ixxie** [reply]
Disappointing, but thank you for the response nonetheless!

**Comment by jheer** [comment]
As @domoritz describes above, Mosaic _as infrastructure_ is not the right level of abstraction for image export. However, _vgplot_ as a library (that happens to be implemented atop Mosaic) could support some level of export. It wouldn't be too hard to add SVG export for a `Plot` instance (hint: PRs welcome!). However, legends would not initially be included as they are actually HTML-based (not pure SVG).

A separate approach would be to approach export using tools external to Mosaic. For example, using Puppeteer one could generate PNG or PDF snapshots server-side at the browser level of abstraction. (FWIW this is what we do in [Living Papers](https://github.com/uwdata/living-papers) to create snapshots of arbitrary web content for inclusion in static media.)
